### PR TITLE
[caffe2]seperate mkl, mklml, and mkldnn

### PR DESCRIPTION
**Pull request by gujinghui**
1. Remove avx2 support in mkldnn
2. Seperate mkl, mklml, and mkldnn
3. Fix convfusion test case

### CODE REVIEW

**Comment by gujinghui** [comment]
@yinghai @jgong5 
pls review.

**Review by yinghai — COMMENTED** [review]
What's the test plan for this change? 

And looks like we still have FIndMKLDNN.cmake. Can we merge that into one?

**Comment by gujinghui** [comment]
@yinghai 
The test cases under caffe2/python/ideep/ covers these changes.

For FindMKLDNN.cmake, yes, I have plan to move all to FindMKLDNN.cmake.
After this PR, I will submit a PR to do this ASAP.
Pls merge this PR firstly. Thanks.

**Comment by yinghai** [comment]
> What's the test plan for this change?

Sorry what I meant is what's your command to build pytorch?

**Comment by gujinghui** [comment]
@yinghai 
Pls try below cmd.
cmake -DBLAS=MKL -DUSE_IDEEP=ON ..

**Comment by yinghai** [comment]
@orionr I think we cannot build with `cmake` along anymore? We need `setup.py` to build things properly, right?

**Comment by gujinghui** [comment]
@yinghai 
Is there any blocking issue for this PR?
If any, pls contact us anytime.

**Comment by orionr** [comment]
@yinghai sorry for the slow response here. libtorch builds (for C++ and no Python) are built exclusively with CMake, but PyTorch and Caffe2 builds will always go through `setup.py`. Changes like this should really match with changes in `tools/setup_helpers/mkldnn`. Ultimately when we move `_C` build to CMake this shouldn't be needed, but it's required right now. We also would need to call all `third_party` builds from root CMake rather than `build_pytorch_libs.sh`. So, long story short, you'll want to make sure `tools/setup_helpers/mkldnn` and `setup.py` match these changes. :)

cc @anderspapitto.

**Comment by orionr** [comment]
Also, @gujinghui, to test, run `python setup.py build_deps` rather than calling CMake directly.

**Comment by gujinghui** [comment]
@yinghai @orionr 
please help review. The test cmd is `export NO_MKLDNN=NO` and `python setup.py build_deps`

**Comment by gujinghui** [comment]
@orionr 
Thanks for detail descriptions.
With this PR, the mkldnn is built from source with caffe2 by setup.py.
No need to seek the libmkldnn.so in system environment.

**Comment by gujinghui** [comment]
@yinghai @orionr 
rebased. pls help review.

File: caffe2/core/macros.h.in [hunk 1]
```diff
@@ -76,6 +76,7 @@ static_assert(
   {"USE_EIGEN_FOR_BLAS", "${CAFFE2_USE_EIGEN_FOR_BLAS}"}, \
   {"USE_LITE_PROTO", "${CAFFE2_USE_LITE_PROTO}"}, \
   {"USE_MKL", "${CAFFE2_USE_MKL}"}, \
+  {"USE_MKLDNN", "${CAFFE2_USE_IDEEP}"}, \
```
Comment (orionr) [inline caffe2/core/macros.h.in]: Do you want to change `CAFFE2_USE_IDEEP` to `CAFFE2_USE_MKLDNN`? I'd like these to be consistent.

File: cmake/Modules/FindMKL.cmake [hunk 2]
```diff
@@ -28,6 +21,7 @@ SET(MKL_LAPACK_LIBRARIES)
 SET(MKL_SCALAPACK_LIBRARIES)
 SET(MKL_SOLVER_LIBRARIES)
 SET(MKL_CDFT_LIBRARIES)
+SET(CAFFE2_USE_MKL OFF)
```
Comment (orionr) [inline cmake/Modules/FindMKL.cmake]: Generally these modules match upstream versions and don't have any Caffe2/PyTorch specific code in them... is there a way to have this not set `CAFFE2_USE_MKL` directly?

File: cmake/Modules/FindMKL.cmake [hunk 3]
```diff
@@ -286,7 +279,8 @@ SET(CMAKE_LIBRARY_PATH ${saved_CMAKE_LIBRARY_PATH})
 SET(CMAKE_INCLUDE_PATH ${saved_CMAKE_INCLUDE_PATH})
 IF (MKL_LIBRARIES AND MKL_INCLUDE_DIR)
   SET(MKL_FOUND TRUE)
-  set(MKL_cmake_included true)
+  SET(CAFFE2_USE_MKL ON)
```
Comment (orionr) [inline cmake/Modules/FindMKL.cmake]: Let's rely on MKL_FOUND instead of `CAFFE2_USE_MKL`

Comment (orionr) [reply cmake/Modules/FindMKL.cmake]: Also, does all of this need to get upstreamed to Intel's repo?

File: cmake/Modules/FindMKLDNN.cmake [hunk 4]
```diff
@@ -1,32 +1,116 @@
...
 #
 # The following are set after configuration is done:
-#  MKLDNN_FOUND
-#  MKLDNN_INCLUDE_DIRS
-#  MKLDNN_LIBRARIES
-#  MKLDNN_LIBRARY_DIRS
+#  MKLDNN_FOUND          : set to true if mkl-dnn is found.
+#  MKLDNN_INCLUDE_DIR    : path to mkl-dnn include dir.
+#  MKLDNN_LIBRARIES      : list of libraries for mkl-dnn
+#  CAFFE2_USE_IDEEP      : set to ON if MKLDNN_FOUND is true
```
Comment (orionr) [inline cmake/Modules/FindMKLDNN.cmake]: Same here on the `CAFFE2_*` business.

File: cmake/Summary.cmake [hunk 5]
```diff
@@ -95,11 +95,8 @@ function (caffe2_print_configuration_summary)
     message(STATUS "    LMDB version        : ${LMDB_VERSION}")
   endif()
   message(STATUS "  USE_METAL             : ${USE_METAL}")
-  message(STATUS "  USE_MKL               : ${USE_MKL}")
-  if(${USE_MKL})
-    message(STATUS "    USE_MKLML           : ${USE_MKLML}")
-    message(STATUS "    USE_IDEEP           : ${USE_IDEEP}")
-  endif()
+  message(STATUS "  USE_MKL               : ${CAFFE2_USE_MKL}")
+  message(STATUS "  USE_MKLDNN            : ${CAFFE2_USE_IDEEP}")
```
Comment (orionr) [inline cmake/Summary.cmake]: There should really be a 1-to-1 between the message and variable so let's call these `USE_MKL` and `USE_MKLDNN` if that works.

**Review by orionr — CHANGES_REQUESTED** [review]
Great work - thank you! I think we need to cleanup some of the naming and, most importantly, remove the CAFFE2_* variables from cmake/Modules/. Those should also likely be upstreamed as well. Happy to talk more on the PR, though.

File: cmake/Modules/FindMKL.cmake (hunk 2 above)
Comment (gujinghui) [reply cmake/Modules/FindMKL.cmake]: pls see blow comments.

File: cmake/Modules/FindMKL.cmake (hunk 3 above)
Comment (gujinghui) [reply cmake/Modules/FindMKL.cmake]: pls see blow comments.

**Comment by gujinghui** [comment]
@orionr 
Understood your concern.
But the USE_MKL has been used in mkl-dnn, as well, which should introduce some conflictions.

And USE_MKLDNN/USE_MKL is defined by user to try to enable MKLDNN/MKL.
But, sometimes we cannot find MKLDNN/MKL.
In order to avoid fusion, new macros are needed here.
How about use below macros in both cmake and C code, instead?
CAFFE2_USE_MKL  --> MKL_FOUND
CAFFE2_USE_MKLDNN --> MKLDNN_FOUND

**Comment by yinghai** [comment]
> How about use below macros in both cmake and C code, instead?
> CAFFE2_USE_MKL --> MKL_FOUND
> CAFFE2_USE_MKLDNN --> MKLDNN_FOUND

If this is the case, I feel we should just use `CAFFE2_USE_MKL`, because in macros.in.h we have a bunch of them already.  What I'm not update with the build plan. @orionr, what's your take?

**Comment by gujinghui** [comment]
hi @orionr @yinghai 
any update on this PR?
what else should be changed for this?

**Comment by yinghai** [comment]
@orionr

**Comment by orionr** [comment]
Thanks for the ping.

I think there are still CAFFE2_* variables being set in cmake/Modules/Find*.cmake files? If so, that needs to be fixed before this can land. If these are only going to be used for PyTorch / Caffe2 (and not for other projects) in the future (I don't know the answer to that) we could move them to a new location, but hopefully we can make them standardized.

**Comment by gujinghui** [comment]
@orionr 
Thanks for reply.
All CAFFE2_* var have been moved out of cmake/Modules/FindMKL*.cmake.

@yinghai @orionr 
Pls help review it.

Thanks,
Jinghui

File: cmake/Modules/FindMKLDNN.cmake [hunk 6]
```diff
@@ -1,32 +1,115 @@
...
-  set(MKLDNN_INCLUDE_DIRS ${MKLDNN_INCLUDE_DIR})
-  set(MKLDNN_LIBRARIES ${MKLDNN_LIBRARY})
-  message(STATUS "Found MKLDNN      (include: ${MKLDNN_INCLUDE_DIR}, library: ${MKLDNN_LIBRARY})")
-  mark_as_advanced(MKLDNN_ROOT_DIR MKLDNN_LIBRARY MKLDNN_INCLUDE_DIR)
-endif()
+FIND_PACKAGE(BLAS)
+FIND_PATH(IDEEP_INCLUDE_DIR ideep.hpp PATHS ${IDEEP_ROOT} PATH_SUFFIXES include)
+FIND_PATH(MKLDNN_INCLUDE_DIR mkldnn.hpp mkldnn.h PATHS ${MKLDNN_ROOT} PATH_SUFFIXES include)
+IF (NOT MKLDNN_INCLUDE_DIR)
+  EXECUTE_PROCESS(COMMAND git submodule update --init mkl-dnn WORKING_DIRECTORY ${IDEEP_ROOT})
```
Comment (orionr) [inline cmake/Modules/FindMKLDNN.cmake]: Would this have any problem on Windows?

**Review by orionr — COMMENTED** [review]
Looks great! Thank you. One question on Windows and will now look at the CI results.

**Comment by orionr** [comment]
Looks like we have some CI failures with this - let's make sure those pass before importing, since this is a pretty significant change. Closing and reopening to see if we can get them restarted.

File: cmake/Modules/FindMKLDNN.cmake (hunk 6 above)
Comment (gujinghui) [reply cmake/Modules/FindMKLDNN.cmake]: Do you mean the different path separator on windows and linux? It should be OK.

**Comment by gujinghui** [comment]
@orionr 
Looks like the rest failures are not caused by this PR.
1. pytorch-win-ws2016-cuda9-cudnn7-py3-build:
> 00:56:41 ..\caffe2\core\nomnigraph\include\nomnigraph/Representations/NeuralNet.h(456): error C2491: 'nom::repr::nn::createOperator': definition of dllimport function not allowed

2. caffe2_py2_gcc4_8_ubuntu14_04_test 
> Oct 24 11:46:01 =================================== FAILURES ===================================
Oct 24 11:46:01 _____________________________ TestGlu.test_glu_old _____________________________
Oct 24 11:46:01 
Oct 24 11:46:01 self = <caffe2.python.operator_test.glu_op_test.TestGlu testMethod=test_glu_old>
Oct 24 11:46:01 args = (), kwargs = {}
Oct 24 11:46:01 
Oct 24 11:46:01     def func(self, *args, **kwargs):
Oct 24 11:46:01 >       self.should_serialize = True
Oct 24 11:46:01 
Oct 24 11:46:01 /usr/local/caffe2/lib/python2.7/dist-packages/caffe2/python/serialized_test/serialized_test_util.py:34: 
Oct 24 11:46:01 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Oct 24 11:46:01 /usr/local/lib/python2.7/dist-packages/hypothesis/_settings.py:248: in new_test
Oct 24 11:46:01     test(*args, **kwargs)
Oct 24 11:46:01 /usr/local/caffe2/lib/python2.7/dist-packages/caffe2/python/serialized_test/serialized_test_util.py:37: in func
Oct 24 11:46:01     hyp_func(self, *args, **kwargs)
Oct 24 11:46:01 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Oct 24 11:46:01 
Oct 24 11:46:01 self = <caffe2.python.operator_test.glu_op_test.TestGlu testMethod=test_glu_old>
Oct 24 11:46:01 
Oct 24 11:46:01     @settings(suppress_health_check=[HealthCheck.filter_too_much])
Oct 24 11:46:01 >   @serial.given(
Oct 24 11:46:01         X=hu.tensor(),
Oct 24 11:46:01         axis=st.integers(min_value=0, max_value=3),
Oct 24 11:46:01         **hu.gcs
Oct 24 11:46:01     )
Oct 24 11:46:01     def test_glu_old(self, X, axis, gc, dc):
Oct 24 11:46:01 E   FailedHealthCheck: It looks like your strategy is filtering out a lot of data. Health check found 50 filtered examples but only 9 good ones. This will make your tests much slower, and also will probably distort the data generation quite a lot. You should adapt your strategy to filter less. This can also be caused by a low max_leaves parameter in recursive() calls

3. pytorch_linux_trusty_py3_6_gcc7_build_test
> Oct 24 10:23:25 ======================================================================
Oct 24 10:23:25 FAIL: test_all_reduce_product (__main__.TestDistBackend)
Oct 24 10:23:25 ----------------------------------------------------------------------
Oct 24 10:23:25 Traceback (most recent call last):
Oct 24 10:23:25   File "test_distributed.py", line 1288, in wrapper
Oct 24 10:23:25     self._join_and_reduce(fn)
Oct 24 10:23:25   File "test_distributed.py", line 1371, in _join_and_reduce
Oct 24 10:23:25     self.assertEqual(p.exitcode, first_process.exitcode)
Oct 24 10:23:25   File "/var/lib/jenkins/workspace/test/common_utils.py", line 404, in assertEqual
Oct 24 10:23:25     super(TestCase, self).assertEqual(x, y, message)
Oct 24 10:23:25 AssertionError: None != 1 : 
Oct 24 10:23:25 
Oct 24 10:23:25 ----------------------------------------------------------------------
Oct 24 10:23:25 Ran 62 tests in 611.141s
Oct 24 10:23:25 
Oct 24 10:23:25 FAILED (failures=1, skipped=30)
Oct 24 10:23:25 Traceback (most recent call last):
Oct 24 10:23:25   File "test/run_test.py", line 394, in <module>
Oct 24 10:23:25     main()
Oct 24 10:23:25   File "test/run_test.py", line 386, in main
Oct 24 10:23:25     raise RuntimeError(message)
Oct 24 10:23:25 RuntimeError: test_distributed failed!
Oct 24 10:23:25 + cleanup
Oct 24 10:23:25 + retcode=1
Oct 24 10:23:25 + set +x
Oct 24 10:23:25 =================== sccache compilation log ===================
Oct 24 10:23:25 ERROR:sccache::server: ["null"] fatal error: Permission denied (os error 13)
Oct 24 10:23:25 
Oct 24 10:23:25 ERROR:sccache::server: ["null"] 	Permission denied (os error 13)

**Comment by ezyang** [comment]
GLU health should be fixed on master.

File: cmake/Modules/FindMKLDNN.cmake (hunk 6 above)
Comment (orionr) [reply cmake/Modules/FindMKLDNN.cmake]: Also that git might need to be `git.exe` or something in the non-Cygwin case. Maybe just wrap this in an `IF (UNIX)`?

**Comment by orionr** [comment]
@gujinghui can you please rebase and force push to this branch so we can confirm all the tests pass? Thanks.

**Review by orionr — APPROVED** [review]

**Comment by orionr** [comment]
Thank you! Failing test is only a timeout, so CI looks good. Starting to land.

**Pull request closed without merging**